janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/computer/FulgoraGraphComputer.java

```diff
@@ -162,193 +179,241 @@ public GraphComputer mapReduce(final MapReduce mapReduce) {
+                "[" + iteration + "]. Computer is aborting.");
+        }
+        return jobResult;
+    }
+
+    private void executeOnPartitionedVertices(int iteration, PartitionedVertexProgramExecutor programExecutor, ScanMetrics jobResult) {
+        programExecutor.run(numThreads, jobResult);
+        long failures = jobResult.getCustom(PartitionedVertexProgramExecutor.PARTITION_VERTEX_POSTFAIL);
+        if (failures > 0) {
+            throw new JanusGraphException("Failed to process [" + failures + "] partitioned vertices in vertex " +
+                "program iteration [" + iteration + "]. Computer is aborting.");
+        }
+    }
+
+    private Map<MapReduce, FulgoraMapEmitter> collectMapJobs() {
+        Map<MapReduce, FulgoraMapEmitter> mapJobs = new HashMap<>(mapReduces.size());
+        for (MapReduce mapReduce : mapReduces) {
+            if (mapReduce.doStage(MapReduce.Stage.MAP)) {
+                FulgoraMapEmitter mapEmitter = new FulgoraMapEmitter<>(mapReduce.doStage(MapReduce.Stage.REDUCE));
+                mapJobs.put(mapReduce, mapEmitter);
             }
-            // Execute map jobs
-            jobId = name + "#map";
-            try (VertexMapJob.Executor job = VertexMapJob.getVertexMapJob(graph, vertexMemory, mapJobs)) {
-                StandardScanner.Builder scanBuilder = graph.getBackend().buildEdgeScanJob();
-                scanBuilder.setJobId(jobId);
-                scanBuilder.setNumProcessingThreads(numThreads);
-                scanBuilder.setWorkBlockSize(readBatchSize);
-                scanBuilder.setJob(job);
-                try {
-                    ScanMetrics jobResult = scanBuilder.execute().get();
-                    long failures = jobResult.get(ScanMetrics.Metric.FAILURE);
-                    if (failures > 0) {
-                        throw new JanusGraphException("Failed to process [" + failures + "] vertices in map phase. Computer is aborting.");
-                    }
-                    failures = jobResult.getCustom(VertexMapJob.MAP_JOB_FAILURE);
-                    if (failures > 0) {
-                        throw new JanusGraphException("Failed to process [" + failures + "] individual map jobs. Computer is aborting.");
+        }
+        return mapJobs;
+    }
+    private void executeMapJobs(Map<MapReduce, FulgoraMapEmitter> mapJobs) {
+        jobId = name + "#map";
+        try (VertexMapJob.Executor job = VertexMapJob.getVertexMapJob(graph, vertexMemory, mapJobs)) {
+            executeMapJob(job);
+            executeReducePhase(mapJobs);
+        }
+        memory.attachReferenceElements(graph);
+    }
+
+    private void executeMapJob(VertexMapJob.Executor job) {
+        StandardScanner.Builder scanBuilder = graph.getBackend().buildEdgeScanJob();
+        scanBuilder.setJobId(jobId);
+        scanBuilder.setNumProcessingThreads(numThreads);
+        scanBuilder.setWorkBlockSize(readBatchSize);
+        scanBuilder.setJob(job);
+        try {
```

Comment:
I do understand that this part of the code was in the old `FulgoraGraphComputer` but I am not sure if we need to wrap `JanusGraphException` in `JanusGraphException` here? Is there any reason for this? I would suggest to add one more `catch` clause and throw unwrapped exception. Something like:
```
} catch (JanusGraphException e) {
    throw e;
} catch (Exception e) {
    throw new JanusGraphException(e);
}
```